# **Extração dos Ficheiros**

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/shahpranshu27/rossman-store-sales/rossmann-store-sales/sample_submission.csv
/kaggle/input/datasets/shahpranshu27/rossman-store-sales/rossmann-store-sales/store.csv
/kaggle/input/datasets/shahpranshu27/rossman-store-sales/rossmann-store-sales/train.csv
/kaggle/input/datasets/shahpranshu27/rossman-store-sales/rossmann-store-sales/test.csv


# **Milestone 1 - Exploração do Dataset**

In [2]:
import pandas as pd

base = "/kaggle/input/datasets/shahpranshu27/rossman-store-sales/rossmann-store-sales"

train = pd.read_csv(f"{base}/train.csv", dtype={"StateHoliday": str}, low_memory=False)
store = pd.read_csv(f"{base}/store.csv")

print("=== Dados Treino ===")
display(train.head())
print(' ')
print("\n=== Dados Loja ===")
display(store.head())

=== Dados Treino ===


,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday
0,1,5,2015-07-31,5263,555,1,1,0,1
1,2,5,2015-07-31,6064,625,1,1,0,1
2,3,5,2015-07-31,8314,821,1,1,0,1
3,4,5,2015-07-31,13995,1498,1,1,0,1
4,5,5,2015-07-31,4822,559,1,1,0,1


 

=== Dados Loja ===


,Store,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [3]:
#Analise do Dataset
print("=== TRAIN ===")
print('Informação do Dataset')
train.info()
print(' ')
print('Descrição do Dataset')
train.describe()


=== TRAIN ===
Informação do Dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB
 
Descrição do Dataset


,Store,DayOfWeek,Sales,Customers,Open,Promo,SchoolHoliday
count,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06,1.017209e+06
mean,5.584297e+02,3.998341e+00,5.773819e+03,6.331459e+02,8.301067e-01,3.815145e-01,1.786467e-01
std,3.219087e+02,1.997391e+00,3.849926e+03,4.644117e+02,3.755392e-01,4.857586e-01,3.830564e-01
min,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.800000e+02,2.000000e+00,3.727000e+03,4.050000e+02,1.000000e+00,0.000000e+00,0.000000e+00
50%,5.580000e+02,4.000000e+00,5.744000e+03,6.090000e+02,1.000000e+00,0.000000e+00,0.000000e+00
75%,8.380000e+02,6.000000e+00,7.856000e+03,8.370000e+02,1.000000e+00,1.000000e+00,0.000000e+00
max,1.115000e+03,7.000000e+00,4.155100e+04,7.388000e+03,1.000000e+00,1.000000e+00,1.000000e+00


In [4]:
print("=== TRAIN ===")
print(train.isnull().sum())

print("\n=== STORE ===")
print(store.isnull().sum())

=== TRAIN ===
Store            0
DayOfWeek        0
Date             0
Sales            0
Customers        0
Open             0
Promo            0
StateHoliday     0
SchoolHoliday    0
dtype: int64

=== STORE ===
Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64


## Análise de Valores em Falta (Missing Values)

A verificação de nulos foi feita separadamente nas duas tabelas, por terem naturezas distintas.

**`train.csv`**: Não foram detetados valores em falta em nenhuma das 9 colunas 
(1 017 209 registos completos). Tratando-se de um histórico de vendas diárias, os dados 
estão íntegros, não sendo necessária qualquer imputação.

**`store.csv`**: Os valores em falta concentram-se nesta tabela e, ao contrário do 
habitual, não correspondem a erros de recolha, mas sim a ausência de evento (nulos 
com significado). A estratégia foi, por isso, definida coluna a coluna:

- **`CompetitionDistance`** (3 nulos): a ausência sugere que não há concorrente registado 
  nas proximidades. Será imputada pela mediana, por ser robusta a valores extremos 
  (há lojas com concorrentes muito distantes que enviesariam a média).
- **`CompetitionOpenSinceMonth` / `CompetitionOpenSinceYear`** (354 nulos cada): indicam 
  que se desconhece a data de abertura do concorrente. Serão preenchidas com 0, 
  representando "informação indisponível".
- **`Promo2SinceWeek` / `Promo2SinceYear` / `PromoInterval`** (544 nulos cada): estes 
  nulos ocorrem exatamente nas 544 lojas onde `Promo2 = 0`, ou seja, lojas que não 
  aderiram à promoção contínua. O nulo significa "não aplicável", pelo que serão 
  preenchidos com 0 (e a `PromoInterval` com a categoria "Nenhum"), preservando 
  essa informação.

**Decisão metodológica:** optou-se por não imputar indiscriminadamente pela média/mediana, 
porque na maioria dos casos o nulo é informação relevante. Imputar cegamente apagaria o 
sinal de que uma loja não tem concorrência próxima ou não participa numa promoção contínua.

In [5]:
# Confirmar o intervalo temporal real dos dados
train["Date"] = pd.to_datetime(train["Date"])

print("Data mais antiga:", train["Date"].min().date())
print("Data mais recente:", train["Date"].max().date())
print("Total de dias distintos:", train["Date"].nunique())

Data mais antiga: 2013-01-01
Data mais recente: 2015-07-31
Total de dias distintos: 942


In [6]:
# Conversão da coluna Date para o tipo datetime (estava como texto)
train["Date"] = pd.to_datetime(train["Date"])

print("Tipo da coluna Date:", train["Date"].dtype)
print("Período coberto:", train["Date"].min().date(), "a", train["Date"].max().date())

Tipo da coluna Date: datetime64[ns]
Período coberto: 2013-01-01 a 2015-07-31


---
## Resumo da Análise Inicial do Dataset (Encerramento Milestone 1)

A inspeção preliminar realizada nesta fase de Data Understanding permite consolidar as observações fundamentais que orientarão o desenvolvimento subsequente do projeto.

### Observações estruturais

| Aspeto | Constatação | Implicação |
| :--- | :--- | :--- |
| Volume | 1 017 209 registos de vendas diárias (train) e 1 115 lojas (store) | Volume amplo, adequado a algoritmos de Machine Learning robustos. |
| Dimensionalidade | 9 colunas no train e 10 no store, combináveis por Store | Estrutura relacional simples, exige um merge entre as duas tabelas. |
| Integridade | Train sem nulos; nulos concentrados no store | Imputação necessária apenas no store, com estratégia por coluna. |
| Tipologia | Mistura de numéricas e categóricas (StoreType, Assortment, StateHoliday) | Exige encoding das categóricas antes da modelação. |
| Natureza temporal | A coluna Date cobre uma série diária por loja | Favorece variáveis de sazonalidade e impõe uma divisão treino e teste temporal. |

### Observações estatísticas relevantes

A variável alvo Sales apresenta forte assimetria positiva, com mínimo em 0 e máximo a rondar os 41 551, o que sugere testar uma transformação logarítmica na modelação.

Cerca de 17% dos registos correspondem a dias de loja fechada (Open igual a 0), nos quais as vendas são sempre nulas. Estes registos serão filtrados no treino do modelo de vendas.

A variável Customers correlaciona-se fortemente com as vendas, mas não está disponível no conjunto de teste real, pelo que será usada apenas na análise exploratória e não como preditor.

Nos atributos de loja, os nulos não correspondem a erros mas sim a ausência de evento, como lojas sem concorrente próximo ou sem adesão à promoção contínua Promo2.

### Conclusões estratégicas para as fases seguintes

1. Métricas: tratando-se de uma regressão, as métricas centrais serão o RMSPE e o MAE, descartando qualquer métrica de classificação.
2. Algoritmos: modelos baseados em árvores como Random Forest e XGBoost são preferíveis aos lineares, pela capacidade de captar relações não lineares e interações.
3. Divisão treino e teste: será feita de forma temporal, reservando as semanas finais para teste, de modo a evitar fugas de informação do futuro para o passado.
4. Engenharia de atributos: prevê-se a criação de variáveis de calendário e de histórico de vendas, a validar quanto à correlação com o alvo na Milestone 2.

O projeto avança agora para a fase de Análise Exploratória de Dados (Milestone 2) com um plano técnico claro e fundamentado.